timeboost-sequencer/src/decrypt.rs

```diff
@@ -282,272 +208,318 @@ impl Worker {
+    }
+
+    /// scan through the inclusion list and extract the relevant ciphertext from encrypted bundle/tx, preserving the order,
+    /// "relevant" means encrypted under the keyset this worker belongs to.
+    fn extract_ciphertexts(&self, incl: &InclusionList) -> Result<Vec<Ciphertext>> {
+        let kid = self.committee.id();
+        let cts = incl
+            .priority_bundles()
+            .iter()
+            .filter(|pb| pb.bundle().kid() == Some(kid))
+            .map(|pb| pb.bundle().data())
+            .chain(
+                incl.regular_bundles()
+                    .iter()
+                    .filter(|b| b.kid() == Some(kid))
+                    .map(|b| b.data()),
+            )
+            .map(|bytes| deserialize::<Ciphertext>(bytes))
+            .collect::<Result<Vec<_>>>()?;
+        Ok(cts)
+    }
+
+    /// Produce decryption shares for each *relevant* encrypted bundles inside the inclusion list,
+    /// where "relevant" means targetted to the same `KeysetId` as the current decrypter/worker.
+    /// Also see [`DecShareBatch`] doc.
+    async fn decrypt(&mut self, round: RoundNumber, incl: InclusionList) -> Result<DecShareBatch> {
+        let ciphertexts = self.extract_ciphertexts(&incl)?;
+
+        let mut dec_shares = vec![];
+        for ct in ciphertexts {
             let dec_share = <DecryptionScheme as ThresholdEncScheme>::decrypt(
                 self.dec_sk.privkey(),
-                &ciphertext,
+                &ct,
                 &vec![],
             )
             .map_err(DecryptError::Decryption)?;
 
-            cids.push(cid);
-            kids.push(*kid);
             dec_shares.push(dec_share);
         }
-
-        Ok(ShareInfo::new(round, kids, cids, dec_shares))
+        Ok(DecShareBatch {
+            round,
+            kid: self.committee.id(),
+            dec_shares,
+        })
     }
 
-    fn insert_shares(&mut self, share_info: ShareInfo) -> Result<()> {
-        let kids = share_info.kids();
-        let cids = share_info.cids();
-        let shares = share_info.dec_shares();
-        // add shares to the incubator using round, keyset id and nonce as key.
-        (0..cids.len()).try_for_each(|i| {
-            let kid = *kids.get(i).ok_or(DecryptError::InvalidMessage)?;
-            let cid = cids.get(i).ok_or(DecryptError::InvalidMessage)?;
-            let share = shares
-                .get(i)
-                .ok_or(DecryptError::InvalidMessage)?
-                .to_owned();
-            let key = DecShareKey::new(share_info.round(), *cid, kid);
-            self.shares
-                .entry(key)
-                .and_modify(|shares| {
-                    shares.insert(share.index(), share.clone());
-                })
-                .or_insert_with(|| [(share.index(), share)].into_iter().collect());
-            Ok(())
-        })
+    /// update local cache of decrypted shares
+    fn insert_shares(&mut self, batch: DecShareBatch) -> Result<()> {
```

Comment:
Even though this returns a `Result`, `insert_shares` does not seem fallible. I wonder though if perhaps some checks should be added, e.g. empty batches could be rejected. Another problem AFAICS is that peers can freely chose their round number and blow up the cache by selecting rounds far ahead of the current one. In Sailfish we require evidence of the previous round, e.g. a QC of the round number. Perhaps we can think of something similar here?

Reply:
very good suggestion, I plan to add `committee: Committee` as a part of the `Decrypter` state (thus `Worker`'s as well) so that they can verify `Evidence` of previous rounds before decrypting. 

but what gap should I allow? I assume we don't need strictly "QC from prev round", but something like "QC from <= X round ago" as a slack condition in case of pipelined timeboost? 
what should the X be?

Reply:
created this ticket to track, will do it in another PR: https://app.asana.com/1/1208976916964769/project/1210099833957576/task/1210279951025506?focus=true